source/_ext/gitstamp.py

```diff
@@ -0,0 +1,139 @@
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+
+import os
+import datetime
+from sphinx import errors
+from github import Github
+
+# Gets the datestamp of the latest commit on the given file
+# Converts the datestamp into something more readable
+# Skips files whose datestamp we can't parse.
+# Expected git datestamp format: 2017-06-07 11:57:38 +1000
+# Output to June 7, 2017
+
+class GHCachedUser:
+    """
+    A user cached from GitHub
+    """
+    def __init__(self, name):
+        self.name = name
+        self.login = None
+        self.avatar_url = None
+
+def page_context_handler(app, pagename, templatename, context, doctree):
+    import git
+    global g
+    if g is None:
+        # We have already errored about this
+        pass
+    fullpagename = pagename
+    docsrc = ''
+    try:
+        docsrc = app.confdir + "/"
+        if docsrc != "/":
+            fullpagename = docsrc + pagename
+    except KeyError:
+        pass
+
+    # Don't barf on "genindex", "search", etc
+    if not os.path.isfile("%s.rst" % fullpagename):
+        return
+
+    try:
+        line = g.log('--pretty=format:%ai,%ae,%an', '-n 1', "%s.rst" % fullpagename).split(',')
+
+        if line == "":
+            # Don't datestamp generated rst's (e.g. imapd.conf.rst)
+            # Ideally want to check their source - lib/imapoptions, etc, but
+            # that involves getting the source/output pair into the extension.
+            return
+
+        updated = line[0][:10]
+        email = line[1]
+        name = line[2]
+        context['gitstamp'] = datetime.datetime.strptime(
+                updated,
+                "%Y-%m-%d"
+            ).strftime(
+                app.config.gitstamp_fmt
+            )
+
+        user = GHCachedUser(name)
+        if gh:
+            # Look in cache first to avoid spamming GitHub
+            if email in gh_user_cache:
+                user = gh_user_cache[email]
+            else:
+                # Search GitHub and retrieve first user with matching email (if any)
+                gh_users = gh.search_users(email)
+                if gh_users.totalCount:
+                    gh_user = next(iter(gh_users))
+                    user.name = gh_user.name
+                    user.login = gh_user.login
+                    user.avatar_url = gh_user.avatar_url
+                gh_user_cache[email] = user
+        context['gitauthor'] = user.name
+        context['gitlogin'] = user.login
+        context['gitavatar'] = user.avatar_url
```

Comment:
Suggestion: Show a default gitavatar if the user does not have one.

Reply:
Nice! All GitHub users will have an avatar (custom or generated) but I added a placeholder image for other users.